Fix part of #1366: Card view for creator dashboard

526avijitgupta opened the pull request:
Tasks to be completed as a part of this PR:
- [x] Create card view for explorations and add options to switch between card view and list view.
- [x] Add sharing links to card view (Should fix https://github.com/oppia/oppia/issues/2181)

UPDATE:
To be tackled in a separate PR (see #2188), thus removing from the tasks list of this PR.
- [x] ~~Verify and make card view to be compatible with mobile as well~~

526avijitgupta commented:
@wxyxinyu @AllanYangZhou I'm attaching a screenshot of the dashboard with card view:

<img width="1280" alt="screenshot 2016-06-26 05 31 44" src="https://cloud.githubusercontent.com/assets/7583815/16359835/5ad7c820-3b62-11e6-893a-7f0f50a9231c.png">

While testing this manually, I could identify an issue where toggling back to the list view from card view, a delay of around 2 sec can bee observed for the list view to be hidden -- both the list view and card views are visible for a short period of time, before the card view is hidden. A quick google search me to this [SO post](http://stackoverflow.com/questions/26938021/angular-ng-if-or-ng-show-responds-slow-2second-delay) . And even though the first answer works for me, I'm not sure if using that would have any side-effects on any other animations. 

Any ideas as to how can I resolve this issue cleanly ?

wxyxinyu commented:
UI comments:
- when I have an untitled exploration, the card view doesn't show 'Untitled'
![image](https://cloud.githubusercontent.com/assets/5061134/16410791/fbc6900a-3cd8-11e6-82b1-1a08252441ad.png)
In the collection tab:
![image](https://cloud.githubusercontent.com/assets/5061134/16410974/ca5a82c8-3cd9-11e6-83f3-3a36201bd341.png)
- I think the "collection" flag (bottom right of the smiley face) is unnecessary, users know from the tabs above that this is a collection.
- Created and last updated don't seem to have updated correctly.
- We might want to consider having a similar list/tile view for collections as for individual explorations. (Let's discuss this when we meet.)

wxyxinyu commented:
Code looks fine to me, @AllanYangZhou PTAL.

526avijitgupta commented:
@wxyxinyu Inline Replies

> when I have an untitled exploration, the card view doesn't show 'Untitled' 

* Fixed.

> I think the "collection" flag (bottom right of the smiley face) is unnecessary, users know from the tabs above that this is a collection.

* Removed collection flag.

> Created and last updated don't seem to have updated correctly.

* Can you please elaborate on this one ?

> I think it will be nice to add border-radius: 2px; to the css (make the corners slightly rounded, for the top left and bottom left of the left button, and the top right and bottom right of the right button).

* Added border-radius of 5px. (2px didn't seem to make any difference).

> Also suggest a lighter color for the border. Maybe #ddd

* Done

> The "Creator Dashboard" heading is the wrong font. Add font-family: "Capriola", "Roboto", Arial, sans-serif; to the css.

* Done

Also, I think you missed my [initial comment](https://github.com/oppia/oppia/pull/2170#issuecomment-228577785).

wxyxinyu commented:
@526avijitgupta re: the first comment, sorry about missing it! I think the root cause of the slowness is actually because you put ng-if on a component that also has ng-repeat, so it attempts to hide every element of the list, which is slow. I fixed the issue by putting another div (`<div ng-if="myExplorationsView === 'card' && explorationsList.length > 0">`) wrapping the entire card view.

wxyxinyu commented:
Re: the collections, if you look at the screenshot, the created and last updated field both say "null", which seems like a bug.

526avijitgupta commented:
@wxyxinyu Thanks, your fix worked.

About collections, it looks like it was an outstanding bug, since I haven't altered any code related to collections as yet!

wxyxinyu commented:
Hmm, could you look into fixing that as well then? That's quite a big bug.

That can be done in a separate PR, but I'd like @AllanYangZhou to take a look at this PR before merging.

526avijitgupta commented:
Yes, I'll do that in a separate PR. But before that, I will include sharing links at least for the card view.

Also, I believe that including sharing links in the list view doesn't seem viable due to the lack of space.
(The view will become too cluttered if we decide to add a field dedicated to sharing links!)
What do you think ?

526avijitgupta commented:
@AllanYangZhou PTAL whenever free! I've completed the tasks for this PR!
Also, I'm posting a screenshot of how the card view looks right now:

<img width="1280" alt="screenshot 2016-07-01 00 56 38" src="https://cloud.githubusercontent.com/assets/7583815/16501358/b36f8a28-3f26-11e6-943f-1f2aaa168f5b.png">

seanlip commented:
Hi @AllanYangZhou can you PTAL? Would be nice to merge this before this weekend's release.

Thanks!

AllanYangZhou commented:
Hi Avijit, sorry for the slow response.

When I switch to the card view I get the following error: https://docs.angularjs.org/error/$injector/unpr?p0=ExplorationEmbedButtonServiceProvider%20%3C-%20ExplorationEmbedButtonService. Also, when I click on the card nothing happens. If I understand correctly it's supposed to be a link to the editor?

seanlip commented:
Hi @526avijitgupta -- on that note, your changes to the creator dashboard are going out in the release this weekend, so please make sure to test and ensure that everything really works properly. It's the logged-in user's homepage, so it's not great if it breaks! Thanks :-)

AllanYangZhou commented:
Also, I think you are going to run into this error: https://github.com/oppia/oppia/blob/d76f7c591af94319bb67b574d8533d15d526f7a1/core/templates/dev/head/services/explorationContextService.js#L96

526avijitgupta commented:
@AllanYangZhou - It looks like I was using an outdated version of sharing links directive, as you correctly pointed out. I've reverted the previous implementation and made the new changes after merging in upstream/develop . 
@seanlip - I've tested it manually, works fine.

Also, I'm posting a screenshot of the updated UI:
<img width="676" alt="screenshot 2016-07-03 15 53 19" src="https://cloud.githubusercontent.com/assets/7583815/16544949/4e55e3b4-4136-11e6-8892-0cdba7c55230.png">

526avijitgupta commented:
Accommodated 4 cards per row. 3 per row didn't look as good.
Updated UI:

<img width="1280" alt="screenshot 2016-07-03 18 55 31" src="https://cloud.githubusercontent.com/assets/7583815/16545692/d6416ca8-414f-11e6-9e5a-6ba2d6b23e9a.png">

wxyxinyu commented on core/templates/dev/head/dashboard/dashboard.html:
```diff
@@ -77,54 +101,125 @@ <h2 style="margin-bottom: -20px; margin-left: 52px;">Creator Dashboard</h2>
+              <td colspan="3"
+                  ng-if="exploration.status === 'private'"
+                  class="exp-private-text">
+                This exploration is private. Publish to receive statistics.
```
This should be translated.

wxyxinyu commented:
The total width of the cards should be equal to the top bar. But this + the translations can be fixed next time, and I'm merging for now.

wxyxinyu commented:
oh, you also need to add a portugese translation for the "last updated" tooltip.

526avijitgupta commented:
@wxyxinyu Will open up a new PR containing all these changes soon. Thanks!
